### DISCUSSION TITLE
Compile Vaultwarden on RaspberryPi4

### DISCUSSION THREAD

**Discussion by Dommschwenker** [dani-garcia/vaultwarden#4460] [Q&A]
Hello everyone,

I have a problem to compile the Vaultwarden (bin) on my Raspberry Pi4.
After more than an hour it crashes with an error message.

pi@raspberrypi4:~/vaultwarden $ cargo build --features sqlite,postgresql,mysql --release
Compiling vaultwarden v1.0.0 (/home/pi/vaultwarden)
LLVM ERROR: out of memory
Allocation failed
error: could not compile vaultwarden (bin "vaultwarden")

Caused by:
  process didn't exit successfully: `/home/pi/.rustup/toolchains/1.76.0-armv7-unknown-linux-gnueabihf/bin/rustc --crate-name vaultwarden --edition=2021 src/main.rs --error-format=json --json=diagnostic-rendered-ansi,artifacts,future-incompat --diagnostic-width=80 --crate-type bin --emit=dep-info,link -C opt-level=3 -C lto=fat -C codegen-units=1 '--deny=clippy::zero_sized_map_values' '--deny=clippy::verbose_file_reads' --deny=unused_lifetimes --deny=unused_import_braces '--deny=clippy::unused_async' --deny=unused --forbid=unsafe_code '--deny=clippy::unnecessary_self_imports' '--deny=clippy::unnecessary_join' --deny=trivial_numeric_casts --deny=trivial_casts '--deny=clippy::string_to_string' '--deny=clippy::string_add_assign' --deny=rust_2021_compatibility --deny=rust_2018_idioms --deny=pointer_structural_match --deny=noop_method_call --forbid=non_ascii_idents '--deny=clippy::needless_lifetimes' '--deny=clippy::mem_forget' '--deny=clippy::match_wildcard_for_single_variants' '--deny=clippy::manual_string_new' '--deny=clippy::manual_instant_elapsed' '--deny=clippy::manual_assert' '--deny=clippy::macro_use_imports' '--deny=clippy::linkedlist' '--deny=clippy::iter_on_single_items' '--deny=clippy::iter_on_empty_collections' '--deny=clippy::inefficient_to_string' --deny=future_incompatible '--deny=clippy::float_cmp_const' '--deny=clippy::equatable_if_let' --deny=deprecated_in_future '--deny=clippy::clone_on_ref_ptr' '--deny=clippy::cast_lossless' '--allow=clippy::blocks_in_conditions' --cfg 'feature="libsqlite3-sys"' --cfg 'feature="mysql"' --cfg 'feature="postgresql"' --cfg 'feature="sqlite"' -C metadata=6ed2d35059f0bd87 -C extra-filename=-6ed2d35059f0bd87 --out-dir /home/pi/vaultwarden/target/release/deps -C strip=debuginfo -L dependency=/home/pi/vaultwarden/target/release/deps --extern argon2=/home/pi/vaultwarden/target/release/deps/libargon2-3770b66736ff4cd6.rlib --extern bigdecimal=/home/pi/vaultwarden/target/release/deps/libbigdecimal-ec62afa7b6104aa0.rlib --extern bytes=/home/pi/vaultwarden/target/release/deps/libbytes-94f1a3c93ed0c32a.rlib --extern cached=/home/pi/vaultwarden/target/release/deps/libcached-c5e2b69dc2e80f18.rlib --extern chrono=/home/pi/vaultwarden/target/release/deps/libchrono-5d2c296c361c26e8.rlib --extern chrono_tz=/home/pi/vaultwarden/target/release/deps/libchrono_tz-62605d309964ce3c.rlib --extern cookie=/home/pi/vaultwarden/target/release/deps/libcookie-edee310dc811d789.rlib --extern cookie_store=/home/pi/vaultwarden/target/release/deps/libcookie_store-d1eeeeb201c4c4bb.rlib --extern dashmap=/home/pi/vaultwarden/target/release/deps/libdashmap-07d3da5f8b281598.rlib --extern data_encoding=/home/pi/vaultwarden/target/release/deps/libdata_encoding-996c86f1cf6aed5a.rlib --extern data_url=/home/pi/vaultwarden/target/release/deps/libdata_url-acb5b7f5d4261824.rlib --extern diesel=/home/pi/vaultwarden/target/release/deps/libdiesel-4265bca89d84ec64.rlib --extern diesel_migrations=/home/pi/vaultwarden/target/release/deps/libdiesel_migrations-724cffdcd48dd85c.rlib --extern dotenvy=/home/pi/vaultwarden/target/release/deps/libdotenvy-58aaeecd3a81fc27.rlib --extern email_address=/home/pi/vaultwarden/target/release/deps/libemail_address-b6e3563840c151d8.rlib --extern fern=/home/pi/vaultwarden/target/release/deps/libfern-def427e7b7ced8b0.rlib --extern futures=/home/pi/vaultwarden/target/release/deps/libfutures-ae4588a143e19668.rlib --extern governor=/home/pi/vaultwarden/target/release/deps/libgovernor-8f82a8cc14a55e07.rlib --extern handlebars=/home/pi/vaultwarden/target/release/deps/libhandlebars-d380a78c18ddf82b.rlib --extern html5gum=/home/pi/vaultwarden/target/release/deps/libhtml5gum-8b52f007d8815908.rlib --extern job_scheduler_ng=/home/pi/vaultwarden/target/release/deps/libjob_scheduler_ng-4de703f0b840c527.rlib --extern jsonwebtoken=/home/pi/vaultwarden/target/release/deps/libjsonwebtoken-a0e48409c7257fcb.rlib --extern lettre=/home/pi/vaultwarden/target/release/deps/liblettre-322ec307c9c0c351.rlib --extern libsqlite3_sys=/home/pi/vaultwarden/target/release/deps/liblibsqlite3_sys-3dd0f5678ae47108.rlib --extern log=/home/pi/vaultwarden/target/release/deps/liblog-78b5103947b92788.rlib --extern num_derive=/home/pi/vaultwarden/target/release/deps/libnum_derive-616102ff184ea67a.so --extern num_traits=/home/pi/vaultwarden/target/release/deps/libnum_traits-e463929c976f21d8.rlib --extern once_cell=/home/pi/vaultwarden/target/release/deps/libonce_cell-787612454099d8e7.rlib --extern openssl=/home/pi/vaultwarden/target/release/deps/libopenssl-3d148ca8ff208969.rlib --extern paste=/home/pi/vaultwarden/target/release/deps/libpaste-75e4a8fa8de9a53f.so --extern percent_encoding=/home/pi/vaultwarden/target/release/deps/libpercent_encoding-b953e366e7270225.rlib --extern pico_args=/home/pi/vaultwarden/target/release/deps/libpico_args-ac98bb4de7165cad.rlib --extern rand=/home/pi/vaultwarden/target/release/deps/librand-64a8b8907329bd16.rlib --extern regex=/home/pi/vaultwarden/target/release/deps/libregex-e2cb55048f16d6b4.rlib --extern reqwest=/home/pi/vaultwarden/target/release/deps/libreqwest-ddbc8b4034281575.rlib --extern ring=/home/pi/vaultwarden/target/release/deps/libring-409648dd416428d8.rlib --extern rmpv=/home/pi/vaultwarden/target/release/deps/librmpv-4dc98f8db326795f.rlib --extern rocket=/home/pi/vaultwarden/target/release/deps/librocket-3fec35f909957519.rlib --extern rocket_ws=/home/pi/vaultwarden/target/release/deps/librocket_ws-4a7267f9e1092933.rlib --extern rpassword=/home/pi/vaultwarden/target/release/deps/librpassword-7607f91f6fe50d30.rlib --extern semver=/home/pi/vaultwarden/target/release/deps/libsemver-6f1d68c6a9467269.rlib --extern serde=/home/pi/vaultwarden/target/release/deps/libserde-0f5c1c52beed6754.rlib --extern serde_json=/home/pi/vaultwarden/target/release/deps/libserde_json-df2190386580814f.rlib --extern syslog=/home/pi/vaultwarden/target/release/deps/libsyslog-be45a1f88f40c9b4.rlib --extern time=/home/pi/vaultwarden/target/release/deps/libtime-40e96931b727302b.rlib --extern tokio=/home/pi/vaultwarden/target/release/deps/libtokio-4bbf838f66ee2ebf.rlib --extern tokio_tungstenite=/home/pi/vaultwarden/target/release/deps/libtokio_tungstenite-745404d716da6873.rlib --extern totp_lite=/home/pi/vaultwarden/target/release/deps/libtotp_lite-7bf486dae6eefdc3.rlib --extern tracing=/home/pi/vaultwarden/target/release/deps/libtracing-cac0167e7de0886d.rlib --extern url=/home/pi/vaultwarden/target/release/deps/liburl-8fd8de1aef5da311.rlib --extern uuid=/home/pi/vaultwarden/target/release/deps/libuuid-34afe24794ed2367.rlib --extern webauthn_rs=/home/pi/vaultwarden/target/release/deps/libwebauthn_rs-179cb6bd9d92c299.rlib --extern which=/home/pi/vaultwarden/target/release/deps/libwhich-275b1c786bb2a9c4.rlib --extern yubico=/home/pi/vaultwarden/target/release/deps/libyubico-bd09ba9e5a0a7861.rlib -L native=/home/pi/vaultwarden/target/release/build/libsqlite3-sys-30127189bc81de38/out -L native=/usr/lib/arm-linux-gnueabihf/ -L native=/usr/lib/arm-linux-gnueabihf -L native=/home/pi/vaultwarden/target/release/build/ring-250e51ec69b89b83/out -L native=/home/pi/vaultwarden/target/release/build/psm-d498d0c14049d6b7/out --cfg sqlite --cfg mysql --cfg postgresql` (signal: 6, SIGABRT: process abort signal)

Hardware: 
- Raspberry Pi 4 - 4GB RAM + 2GB SWAP
- 64GB SD-Card

Software:
- Raspberry PI OS (32bit)
- Linux raspberrypi4 6.6.20+rpt-rpi-v8 #1 SMP PREEMPT Debian 1:6.6.20-1+rpt1 (2024-03-07) aarch64 GNU/Linux
- Nodejs v16.20.2
- NPM v8.19.4
- rustup v1.27.0 (bbb9276d2 2024-03-08) / Toolchains 1.76.0-armv7-unknown-linux-gnueabihf
- rustc 1.76.0 (07dca489a 2024-02-04)
- cargo 1.76.0 (c84b36747 2024-01-18)

When I use Raspberry Pi OS (64bit) the same error. 
If I compile it on a Linux machine, it works; but this file does not start on the Raspberry Pi (x86_64)

Does anyone have an idea?

Thank you
Dommschwenker

**Comment by BlackDex** [comment]
See #4320

I would suggest to use
https://github.com/dani-garcia/vaultwarden/blob/2d98aa304501b0f710d3a97ec854acdfc115228f/Cargo.toml#L195

**Reply by BlackDex** [reply]
Also, if you have a Pi4 why are you using a 32bit os?

**Reply by Dommschwenker** [reply]
Thanks, it worked with the profile "release-low".

I had previously compiled the vaultwarden(bin) on a RaspberryPi3 under RPi OS 32bit. 
I had also tried under the new RPi OS 64bit, but also without success. 
Now with the profile it now also worked under 64bit.

Thanks for your help
